master/internal/experiment/bulk_action_test.go

```diff
@@ -0,0 +1,532 @@
+package experiment
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/determined-ai/determined/proto/pkg/apiv1"
+)
+
+type bulkActorMock struct {
+	mock.Mock
+}
+
+func (m *bulkActorMock) experimentsEditableByUser(
+	ctx context.Context,
+	projectID int32,
+	experimentIDs []int32,
+	filters *apiv1.BulkExperimentFilters,
+) ([]int32, error) {
+	returns := m.Called(ctx, projectID, experimentIDs, filters)
+	ret0, _ := returns.Get(0).([]int32)
+	return ret0, returns.Error(1)
+}
+
+// experimentMock inherits all the methods of the Experiment
+// interface, but only implements the ones we care about
+// for testing.
+type experimentMock struct {
+	mock.Mock
+	Experiment
+}
+
+func (m *experimentMock) ActivateExperiment() error {
+	returns := m.Called()
+	return returns.Error(0)
+}
+
+func (m *experimentMock) PauseExperiment() error {
+	returns := m.Called()
+	return returns.Error(0)
+}
+
+func (m *experimentMock) CancelExperiment() error {
+	returns := m.Called()
+	return returns.Error(0)
+}
+
+func (m *experimentMock) KillExperiment() error {
+	returns := m.Called()
+	return returns.Error(0)
+}
+
+func TestActivateExperiments(t *testing.T) {
+	type args struct {
+		projectID     int32
+		experimentIds []int32
+		filters       *apiv1.BulkExperimentFilters
+	}
+	type experimentsEditable struct {
+		expIDs []int32
+		err    error
+	}
+
+	ctx := context.Background()
+
+	tests := []struct {
+		name                  string
+		args                  args
+		experimentsEditable   experimentsEditable
+		registeredExperiments []int
+		expectedActivates     []int
+		expectedResults       []ExperimentActionResult
+		expectedErr           bool
+	}{
+		{
+			name:                "user login error",
+			args:                args{},
+			experimentsEditable: experimentsEditable{err: errors.New("user has no permision to edit experiments")},
+			expectedErr:         true,
+		},
+		{
+			name: "three experiments selected, one found",
+			args: args{
+				projectID:     1,
+				experimentIds: []int32{132, 142, 152},
+			},
+			experimentsEditable: experimentsEditable{
+				expIDs: []int32{132, 142},
+			},
+			registeredExperiments: []int{142, 162},
+			expectedActivates:     []int{142},
+			expectedResults: []ExperimentActionResult{
+				{
+					Error: status.Error(codes.NotFound, "experiment '152' not found"),
+					ID:    152,
+				},
+				{
+					Error: status.Errorf(codes.FailedPrecondition, "experiment in terminal state"),
+					ID:    132,
+				},
+				{
+					ID: 142,
+				},
+			},
+		},
+		{
+			name: "filters are used",
```

Comment:
Perhaps more test cases on different filter fields?

Reply:
This has proved trickier than I hoped, and I keep getting pulled off for other issues while trying to make these test cases, which then makes all these tickets bleed into more sprints. I think I'd like to just merge this, and make a new ticket for the additional coverage, if that's OK with you.